internal/queue/queue.go

```diff
@@ -50,40 +80,182 @@ type Queue struct {
 // in the internal kubernetes metrics.
 func New(ratelimiter workqueue.RateLimiter, name string, handler ItemHandler) *Queue {
 	return &Queue{
-		name:      name,
-		workqueue: workqueue.NewNamedRateLimitingQueue(ratelimiter, name),
-		handler:   handler,
+		clock:                    clock.RealClock{},
+		name:                     name,
+		ratelimiter:              ratelimiter,
+		items:                    list.New(),
+		itemsBeingProcessed:      make(map[string]*queueItem),
+		itemsInQueue:             make(map[string]*list.Element),
+		handler:                  handler,
+		wakeupCh:                 make(chan struct{}, 1),
+		waitForNextItemSemaphore: semaphore.NewWeighted(1),
 	}
 }
 
 // Enqueue enqueues the key in a rate limited fashion
-func (q *Queue) Enqueue(key string) {
-	q.workqueue.AddRateLimited(key)
+func (q *Queue) Enqueue(ctx context.Context, key string) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	q.insert(ctx, key, true, 0)
 }
 
 // EnqueueWithoutRateLimit enqueues the key without a rate limit
-func (q *Queue) EnqueueWithoutRateLimit(key string) {
-	q.workqueue.Add(key)
+func (q *Queue) EnqueueWithoutRateLimit(ctx context.Context, key string) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	q.insert(ctx, key, false, 0)
 }
 
 // Forget forgets the key
-func (q *Queue) Forget(key string) {
-	q.workqueue.Forget(key)
+func (q *Queue) Forget(ctx context.Context, key string) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	ctx, span := trace.StartSpan(ctx, "Forget")
+	defer span.End()
+
+	ctx = span.WithFields(ctx, map[string]interface{}{
+		"queue": q.name,
+		"key":   key,
+	})
+
+	if item, ok := q.itemsInQueue[key]; ok {
+		span.WithField(ctx, "status", "itemInQueue")
+		delete(q.itemsInQueue, key)
+		q.items.Remove(item)
```

Comment:
How would it be possible for item to be nil? I don't see how that could get added.

Reply:
Can't items in this map be nil?  IIUC, yes. And the fact existing code doesn't insert nils doesn't mean some future change won't introduce this and it should be an easy thing to do to protect against it here. But won't block because of this, just wanted to highlight the possible panic.

Reply:
items are added to the iteminqueue table by _us_ and the only exposed interface outside of this package is the ability to add a concrete string key.